### PR TITLE
net: ip: 6lo (IPHC) compression rework

### PR DESCRIPTION
**Pull request by alexanderwachter**
This is a rework of the 6lo IPHC. The compression now works inline
on the original buffer instead of allocation a new one.
Additionally DAM_11 (Destination address fully elided) has precedence
over DAM_10 (16 bit compressible) now.

### CODE REVIEW

File: subsys/net/ip/6lo_private.h [hunk 1]
```diff
@@ -81,6 +81,7 @@
 #define NET_6LO_NHC_UDP_PORT_10		0x02
 #define NET_6LO_NHC_UDP_PORT_11		0x03
 
+#define NET_6LO_NCH_UDP_CHECKSUM    0x04
```
Comment (rveerama1) [inline subsys/net/ip/6lo_private.h]: Is it typo?  "NCH", rest of the macros has NHC.

Comment (alexanderwachter) [reply subsys/net/ip/6lo_private.h]: Typo.

**Comment by alexanderwachter** [comment]
net: ip: net_pkt: additional pkt_cursor_advance after cursor operations. (af268141c08c6fa44f0b024c1f9811579c9ac449) is a workaround and should be fixed in an extra PR.
@tbursztyka Maybe you can spot the problem here.

File: subsys/net/ip/6lo_private.h [hunk 2]
```diff
@@ -81,6 +81,7 @@
 #define NET_6LO_NHC_UDP_PORT_10		0x02
 #define NET_6LO_NHC_UDP_PORT_11		0x03
 
+#define NET_6LO_NHC_UDP_CHECKSUM    0x04
```
Comment (jukkar) [inline subsys/net/ip/6lo_private.h]: Please indent the value against other values, also the empty line before could be removed too.

**Review by jukkar — COMMENTED** [review]
LGTM. Just minor nitpick and also the checkpatch issue should be fixed too.

**Comment by alexanderwachter** [comment]
@jukkar I coulnd't find any "beautiful" solution for the checkpatch error so far. Need to think about this a little more.

File: subsys/net/ip/6lo_private.h (hunk 2 above)
Comment (alexanderwachter) [reply subsys/net/ip/6lo_private.h]: Done

File: subsys/net/ip/net_pkt.c [hunk 3]
```diff
@@ -1823,6 +1823,8 @@ bool net_pkt_is_contiguous(struct net_pkt *pkt, size_t size)
 	if (pkt->cursor.buf && pkt->cursor.pos) {
```
Comment (jukkar) [inline subsys/net/ip/net_pkt.c]: There is some garbage at the end of commit message of "net: ip: net_pkt: Add pkt_cursor_advance before contiguous check"
Also there is redundant information in the commit subject, and could be written shorter like this "net: pkt: Add pkt_cursor_advance before contiguous check"

**Review by jukkar — CHANGES_REQUESTED** [review]

File: subsys/net/ip/6lo.c [hunk 4]
```diff
@@ -1474,6 +1474,8 @@ static bool uncompress_IPHC_header(struct net_pkt *pkt,
 		}
 	}
 
+	net_pkt_cursor_init(pkt);
```
Comment (tbursztyka) [inline subsys/net/ip/6lo.c]: you will need to split this in 2: one on subsys/net and one on tests/net

File: subsys/net/ip/6lo.c [hunk 5]
```diff
@@ -4,6 +4,7 @@
 
 /*
  * Copyright (c) 2016 Intel Corporation
+ * Copyright (c) 2019 Alexander Wachter
```
Comment (tbursztyka) [inline subsys/net/ip/6lo.c]: review the commit title, instead of 'net: ip:'  have 'subys: net:'  or just 'net:' apply that on other commit titles.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
first patch is quite big as it is changing the overall buffer access logic. I guess it would require an actual runtime test of a 15.4 network to see if it did not break anything.

**Comment by alexanderwachter** [comment]
Next step is to remove the hacks in the 802.15.4 code that protects the ll addresses. They are obsolete with this PR

File: subsys/net/ip/6lo.c (hunk 5 above)
Comment (jukkar) [reply subsys/net/ip/6lo.c]: > have 'subys: net:'

There should never be "subsys: net: " prefix as the "net: " is already the subsystem so there is no need to add duplicate information.
Depending on context, sometimes "net: ip: ..." makes sense but in this case for 6lo stuff, the prefix should probably be just "net: 6lo: ..." or similar (depending on the stuff we are changing in the commit).

**Comment by jukkar** [comment]
> require an actual runtime test of a 15.4 network to see if it did not break anything

Yes, I agree here. We would need to check this in 802.15.4 and in BT.

File: subsys/net/ip/6lo.c (hunk 5 above)
Comment (tbursztyka) [reply subsys/net/ip/6lo.c]: > There should never be "subsys: net: " prefix as the "net: " is already the subsystem

I have been using that format without any complaints, moreover when other patches touches tests/net, then tests: net: was the prefix, so it differentiates which part in zephyr targeting net domain.

File: subsys/net/ip/6lo.c (hunk 4 above)
Comment (alexanderwachter) [reply subsys/net/ip/6lo.c]: Not sure here. I need to call  net_pkt_cursor_init(pkt) because i have added a new pkt in front.

**Comment by alexanderwachter** [comment]
Indeed this need to be checked. Does anyone have a setup to check?

File: subsys/net/ip/6lo.c (hunk 4 above)
Comment (tbursztyka) [reply subsys/net/ip/6lo.c]: No my point was the patch touches the net subsystem and the test: better have one patch for each.

File: subsys/net/ip/net_pkt.c (hunk 3 above)
Comment (alexanderwachter) [reply subsys/net/ip/net_pkt.c]: Removed.

File: subsys/net/ip/6lo.c (hunk 4 above)
Comment (alexanderwachter) [reply subsys/net/ip/6lo.c]: Done.

**Comment by rlubos** [comment]
@alexanderwachter I did a quick test of 802.15.4 and nRF52840, seems to work fine :+1:

I'll do some more thorough test (15.4 and IPSP) once you're done with decompression.

**Comment by alexanderwachter** [comment]
Reworked the uncompression too now.

**Comment by rlubos** [comment]
Tested the new commits with 802.15.4 and BLE IPSP. 

The 802.15.4 communication works fine (tested with `echo_client`/`echo_server` samples running on two nRF52840_pca10056 boards).

Sadly I cannot say the same about BLE IPSP, it appears to be broken (same behavior observed on current master, so it's likely not related to this PR). I've tried `echo_server` and `ipsp` samples. In both cases, they do not seem to echo back the TCP data (tried `echo_client` from net-tools and telnet).
The `echo_server`, when tested with UDP packets (`echo_client` again) works fine until a 1280 byte datagram is received - it's dropped and ICMP response is sent (I'm not sure though if such large packets are supported by IPSP). 
Just to be clear - I observe the same issues on current master, so I doubt these problems relate to the changes in this PR. Unfortunately, I have no time right now to investigate it any further. Attached wireshark logs in case someone wanted to have a look.

[ble_ipsp_issues.zip](https://github.com/zephyrproject-rtos/zephyr/files/3367450/ble_ipsp_issues.zip)

**Comment by tbursztyka** [comment]
@alexanderwachter will this PR address also not-in-place header compression? (so user would provide a buffer of its own where compressed header would be put). This would solve an issue we have with TCP where the whole tcp packet has to be cloned prior to be sent.

**Comment by alexanderwachter** [comment]
@tbursztyka I had a look at the TCP stack and I don't think we can solve this issue with a separate buffer for compression. The problem is that if we have a separate buffer we can't insert this buffer to the pkt because if we do so we would need to pull the original header. If we don't insert the compressed buffer we need to handle TCP connections different than others in all L2 using 6Lo.
An alternative would be to recover the original header by decompressing it after the data is sent. This could be done by checking the headroom and do an in-place decompression in the buffers headroom.

**Comment by jukkar** [comment]
We clone the TCP packet before sending the clone packet. The original packet is kept until we have received ACK from peer. It does not really matter what the 6lo does with the cloned packet.

**Comment by alexanderwachter** [comment]
@jukkar we try to avoid cloning the packet but only increase the reference count. The problem is that at the moment we would try to compress the packet again. My preferred solution is to check if the packet is compressed already and skip the compression if so.

**Comment by alexanderwachter** [comment]
@tbursztyka I think we should handle the TCP issue in a follow-up PR. Can you create an issue for that and assign us both?

**Comment by tbursztyka** [comment]
> I had a look at the TCP stack and I don't think we can solve this issue with a separate buffer for compression. The problem is that if we have a separate buffer we can't insert this buffer to the pkt because if we do so we would need to pull the original header. If we don't insert the compressed buffer we need to handle TCP connections different than others in all L2 using 6Lo.

Nope, it can be solved as I planned: 6lo compression would put the result in an external buffer and let net_pkt untouched. Have a look at how 15.4 creates frames from net_pkt: it has a dedicated frame_buffer_data buffer where to create the frame. So compressed header could be put there (for the very first frame to be sent) then it would continue on the net_pkt thanks to the cursor etc...

bt l2 would require some modification as well but in the end we could totally scrap TCP net_pkt clone stuff.

File: subsys/net/ip/6lo.c [hunk 6]
```diff
@@ -46,6 +46,74 @@ static inline u8_t get_6co_cid(struct net_icmpv6_nd_opt_6co *opt)
...
+
+	return size;
+}
+
+static const u8_t tf_inline_size_table[] = {4, 3, 1, 0};
+/* The first bit of the index is SAC        |  SAC=0   |  SAC=1   |*/
+static const u8_t sa_inline_size_table[] = {16, 8, 2, 0, 0, 8, 2, 0};
+
+static const u8_t da_inline_size_table[] =
+/* The first bit is M, the second DAC
```
Comment (tbursztyka) [inline subsys/net/ip/6lo.c]: Put this comment before declaring the variable

File: subsys/net/ip/6lo.c [hunk 7]
```diff
@@ -46,6 +46,74 @@ static inline u8_t get_6co_cid(struct net_icmpv6_nd_opt_6co *opt)
...
+ *	| M=0 DAC=0 | M=0 DAC=1 | M=1 DAC=0  | M=1 DAC=1 (DAM always 00)
+ */
+	{16, 8, 2, 0, 0, 8, 2, 0, 16, 6, 4, 1, 6};
```
Comment (tbursztyka) [inline subsys/net/ip/6lo.c]: Move these variable at the top of the file

**Review by tbursztyka — COMMENTED** [review]
minor issue in uncompression rework

File: subsys/net/ip/6lo.c (hunk 6 above)
Comment (alexanderwachter) [reply subsys/net/ip/6lo.c]: This comment should show the borders of the values but actually, it looks strange.
Moved on top.

File: subsys/net/ip/6lo.c (hunk 7 above)
Comment (alexanderwachter) [reply subsys/net/ip/6lo.c]: Done

**Review by tbursztyka — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Review by rveerama1 — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**